[PORT] Gender Neutrality

Bobbanz1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
What this PR basically does is it adds Non-Binary as a possible gender option

Also adds the ability to customize the character even further by allowing you to pick what body type the character is using, so you can play as a Male…defined? character with a Female body, a Nonbinary with a Male body or a Female with a Male body. 

However this only works if the Species isn’t Agendered, such as the Ethereals, in which case the option doesn’t actually appear in the Character Preferences. 

Also one PR in this port removes the division between male and female under-clothing, facial hair and hair, so you can just use whatever under-clothing, facial hair or hair you want to use for your character. 

Ports a modified PR version from Downstream:
* https://github.com/BeeStation/NSV13/pull/2297

Which ports the following PRs from TG:
* https://github.com/tgstation/tgstation/pull/44530 (NSFW because God that comment section is fucking VICIOUS)
* https://github.com/tgstation/tgstation/pull/44555
* https://github.com/tgstation/tgstation/pull/44597
* https://github.com/tgstation/tgstation/pull/56504
* https://github.com/tgstation/tgstation/pull/58785

And the following PR from Citadel Station:
* https://github.com/Citadel-Station-13/Citadel-Station-13/pull/11581 (Most of it)

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Allows players to further customize their characters.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->
<details>
<summary>Screenshots&Videos</summary>

Close enough, all I'm going to get uploaded today it seems, had a 80 MB video showcasing cloning working but that one apparently didn't want to get uploaded so have this short one showcasing the preferences instead.

https://user-images.githubusercontent.com/59128051/215546543-ad56e320-fa62-4721-8fc8-99818bd9cb17.mp4

</details>

## Changelog
:cl:Shaps/Ryll, Wesoda25, Y0SH1M4S73R, Mothblocks, Ghommie
tweak: NT has updated its personnel policies and allowed for employees to register as gender neutral, with standard support for they/them pronouns.
tweak: changed text on gender options to make them grammatically correct and visually pleasing
spellcheck: Wound tending and people holding items in photographs have had their pronoun usage fixed.
add: Added M/F body preferences.
qol: Preferences are no longer dependent on gender.
Code: I converted some prefs input lists to tgui versions
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Bobbanz1 committed:
[PORT] Gender Neutrality

Bobbanz1 committed:
Merge branch 'BeeStation:master' into AttemptNumberWhatever

itsmeow commented on code/modules/client/preferences.dm:
```diff
@@ -1739,16 +1719,22 @@ GLOBAL_LIST_EMPTY(preferences_datums)
+						active_character.underwear = random_underwear(active_character.gender)
+						active_character.undershirt = random_undershirt(active_character.gender)
+						active_character.socks = random_socks()
+						active_character.facial_hair_style = random_facial_hair_style(active_character.gender)
```
is randomizing all this necessary now that it's all generic? (the options exist on both)

Bobbanz1 replied:
I’ll see what happens if I remove it as I’ve heard some complaining downstream that the constant randomization whenever you make a major change to a character is very annoying.

Bobbanz1 replied:
I've removed the randomizaton that happens to these things.

itsmeow commented on code/game/machinery/computer/medical.dm:
```diff
@@ -534,9 +536,9 @@
 				switch(rand(1,6))
 					if(1)
-						R.fields["name"] = random_unique_name(R.fields["sex"],1)
+						R.fields["name"] = random_unique_name(R.fields["gender"],1)
```
does random_unique_name have gender neutral options?

Bobbanz1 replied:
Nope, seems to either give first_names_female or defaults to male names
Although I've noticed there's a first_names global, could create a version that uses that.

Bobbanz1 replied:
Implemented a neutral option

itsmeow commented on code/modules/client/preferences2/character_save.dm:
```diff
@@ -262,6 +258,10 @@
 	features = random_features()
+	if(gender == MALE || gender != FEMALE)
+		features["body_model"] = gender
+	else if(gender == PLURAL)
```
this if else statement makes 0 sense

Bobbanz1 replied:
Removed and replaced it instead with `if(gender)` and now defaults to randomly picking between a male or female body, assuming the species itself has a gender.

Bobbanz1 committed:
Updates

Removes the randomization which happens upon changing character body type or gender.

Implements a neutral option for random_unique_name, don't know if it works.

Also removes the if else statement that makes no sense.

PowerfulBacon commented on code/datums/datacore.dm:
```diff
@@ -251,7 +251,13 @@
 		G.fields["p_stat"]		= "Active"
 		G.fields["m_stat"]		= "Stable"
-		G.fields["sex"]			= H.gender
+		G.fields["gender"]		= H.gender
```
This should be a switch and not assign before overwriting

Bobbanz1 replied:
Done.

PowerfulBacon commented on code/__HELPERS/sanitize_values.dm:
```diff
@@ -41,7 +41,7 @@
 //more specialised stuff
 /// Return `gender` if it is a valid gender, otherwise `default`. No I did not mean to offend you. -qwerty
-/proc/sanitize_gender(gender,neuter=0,plural=0, default="male")
+/proc/sanitize_gender(gender,neuter=0,plural=1, default="male")
```
Why are we changing the default behaviours over passing in plural as the gender?

Bobbanz1 replied:
Because for some reason plural needs to be enabled or it will return a value of default, which in this case is `"male"`

I could alter it so that it checks for `if(MALE, FEMALE, PLURAL)` which then returns the gender?

Bobbanz1 replied:
Reverted to the previous version and then moved the plural thing into the if(MALE, FEMALE)

PowerfulBacon commented on code/__HELPERS/mobs.dm:
```diff
@@ -127,12 +128,14 @@
 			return pick(GLOB.facial_hair_styles_list)
 
-/proc/random_unique_name(gender, attempts_to_find_unique_name=10)
+/proc/random_unique_name(gender, attempts_to_find_unique_name=10, neutral = FALSE)
```
This is kind of ugly, any reason we can't use the byond plural gender?

Bobbanz1 replied:
> This is kind of ugly, any reason we can't use the byond plural gender?

Depends if you want it to always roll a neutral option or merely do so if the gender is plural?

Bobbanz1 replied:
Reverted to the original version and made a plural gender entry that draws a random name from all the first names

PowerfulBacon commented on code/datums/datacore.dm:
```diff
@@ -289,7 +295,13 @@
 		L.fields["rank"] 		= H.mind.assigned_role
 		L.fields["age"]			= H.age
-		L.fields["sex"]			= H.gender
+		L.fields["gender"]		= H.gender
```
Same here

Bobbanz1 replied:
Will need to check the locked records as those looked weird in the port
Instead of assigning `L.fields["gender] = "Other"`
It was assigning 
`L.fields["gender] = H.gender`
 and then overwriting…whatever this is
`G.fields["gender] = "Other"`

PowerfulBacon commented on code/game/machinery/computer/medical.dm:
```diff
@@ -272,12 +272,14 @@
 								return
 							src.active1.fields["fingerprint"] = t1
-					if("sex")
+					if("gender")
```
Can we make this a drop down box instead, is more convenient if there are more than 2 options

Bobbanz1 replied:
Done, made it the same for the security records console.

PowerfulBacon commented on code/modules/client/preferences.dm:
```diff
@@ -1390,11 +1398,7 @@ GLOBAL_LIST_EMPTY(preferences_datums)
-						new_hair_style = input(user, "Choose your character's hair style:", "Character Preference")  as null|anything in GLOB.hair_styles_male_list
-					else
-						new_hair_style = input(user, "Choose your character's hair style:", "Character Preference")  as null|anything in GLOB.hair_styles_female_list
+					var/new_hair_style = input(user, "Choose your character's hair style:", "Character Preference")  as null|anything in GLOB.hair_styles_list
```
Not super sure how I feel about the global hair style lists without any kind of sorting. I find it hard to find hair styles that fit characters at the moment on the mirror, which I believe can use all. It would be nice if there was some categorisation so finding appropriate hair styles for what you want is easier.

Tsar-Salat replied:
Maybe a button that lets you change between female and male lists? You are already implementing Female or Male body, might as well do hair as well.

itsmeow replied:
This is not a problem of separation, it's that there's no icon previews. Leave it, it's out of scope imo

PowerfulBacon replied:
I don't think its out of scope considering this PR is changing the behaviour to make it arguably worse to sort through, leaving it out would be a case of retaining the current behaviour. A preview would be nice (even if out of scope), but filters are better than nothing.

Bobbanz1 replied:
Could make it be something similar to how I did the mirror choice. 

A choice between picking from the Male or Female Hairstyles, although don’t think that wouldn’t exactly fix the problem?

itsmeow replied:
You could change it to use tgui_input_list which is at least searchable

PowerfulBacon replied:
really should have atomised the gender neutrality part out and kept this bit seperate

PowerfulBacon commented on code/modules/client/preferences.dm:
```diff
@@ -1433,32 +1431,18 @@ GLOBAL_LIST_EMPTY(preferences_datums)
-						new_facial_hair_style = input(user, "Choose your character's facial-hair style:", "Character Preference")  as null|anything in GLOB.facial_hair_styles_male_list
-					else
-						new_facial_hair_style = input(user, "Choose your character's facial-hair style:", "Character Preference")  as null|anything in GLOB.facial_hair_styles_female_list
+					var/new_facial_hair_style = input(user, "Choose your character's facial-hair style:", "Character Preference")  as null|anything in GLOB.facial_hair_styles_list
```
Here female characters are allowed facial hair but in other places, like the mirror, they aren't, which is a weird inconsistency

Bobbanz1 replied:
Mirrors now allow all sides of the Cube to engage in facial hair styling

PowerfulBacon commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -528,12 +528,13 @@
 //Used for new human mobs created by cloning/goleming/podding
 /mob/living/carbon/human/proc/set_cloned_appearance()
-	if(gender == MALE)
+	if(dna.features["body_model"] == MALE)
```
This is horribly ugly, is there no variable that stores the body model on the human?

Bobbanz1 replied:
> This is horribly ugly, is there no variable that stores the body model on the human?

I had to find a way to make it work with cloning, the solution I found was the citadel PR which puts it as a feature of the mob.

Bobbanz1 replied:
Also think that moving it out of the features would make some trouble down the road as changelings would probably be easily identifiable if someone were to strip them and they didn’t look like how the original did. 

Also believe that I experimented with other possible methods of making this work through the TG body type PR port which wasn’t compatible with cloning and my attempts at making it work all resulted in failures.

PowerfulBacon commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -663,7 +663,7 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 			var/mutable_appearance/underwear_overlay
 			if(underwear)
-				if(H.dna.species.sexes && H.gender == FEMALE && (underwear.gender == MALE))
+				if(H.dna.species.sexes && H.dna.features["body_model"] == FEMALE && (underwear.gender == MALE))
```
Again here, there are often runtimes with DNA related things

PowerfulBacon commented on code/modules/research/xenobiology/xenobiology.dm:
```diff
@@ -360,12 +360,7 @@
-			else
-				user.gender = MALE
-				user.visible_message("<span class='boldnotice'>[user] suddenly looks more masculine!</span>", "<span class='boldwarning'>You suddenly feel more masculine!</span>")
+			user.set_gender(user.gender == MALE ? FEMALE : MALE, forced = TRUE) //You are doing this to yourself.
```
Why remove the message that indicates something happened?

Bobbanz1 replied:
Was moved to this proc inheritor.
https://github.com/BeeStation/BeeStation-Hornet/blob/ba1f45be26c47a18801337ffacb456856a7d607d/code/modules/mob/living/carbon/carbon.dm#L987

PowerfulBacon commented on code/modules/research/xenobiology/xenobiology.dm:
```diff
@@ -997,13 +992,8 @@
-		L.gender = MALE
-		L.visible_message("<span class='boldnotice'>[L] suddenly looks more masculine!</span>", "<span class='boldwarning'>You suddenly feel more masculine!</span>")
-	L.regenerate_icons()
+	if(!L.set_gender(L.gender == MALE ? FEMALE : MALE))
```
Same here

PowerfulBacon commented on code/modules/mob/living/living.dm:
```diff
@@ -1375,3 +1375,9 @@
 		eminence.selected_mob = src
 		to_chat(eminence, "<span class='brass'>You select [src].</span>")
+
+/mob/living/proc/set_gender(ngender = NEUTER, silent = FALSE, update_icon = TRUE, forced = FALSE)
```
Why does this not work on mobs with a ckey?

Bobbanz1 replied:
Updated to make it work on mobs with Ckeys

PowerfulBacon commented on code/modules/client/preferences.dm:
```diff
@@ -177,7 +177,15 @@ GLOBAL_LIST_EMPTY(preferences_datums)
+					dispGender = "Female"
+				else
+					dispGender = "Other"
+				dat += "<b>Gender:</b> <a href='?_src_=prefs;preference=gender'>[dispGender]</a><BR>"
```
Personally I think that the behaviour should be
- single button by default indicating gender, with ability to pick male, female or other.
- selecting other unlocks 2 additional buttons, 1 for pronoun and 1 for body type
- selecting male or female assigns the appropriate pronoun and bodytype

Bobbanz1 replied:
Moved it so that it’s now only showing the body model preference if the player picks the "Other" gender. 

I think the other stuff would be out of reach for this PR although I can try to figure out a method to do so

PowerfulBacon replied:
As long as the body model preference aligns to gender automatiaclly if you haven't selected other, then that will be fufilled anyway (In other words selecting female should be guaranteed to give you a female body, and same with male).

Bobbanz1 replied:
Think it defaults to the current gender as body_model if it's not selected as "Other"

PowerfulBacon reviewed (changes requested):
Good changes overall

Bobbanz1 committed:
Some stuff in response to review

Bobbanz1 committed:
Merge branch 'master' into AttemptNumberWhatever

Bobbanz1 committed:
Merge remote-tracking branch 'upstream/master' into AttemptNumberWhatever

itsmeow commented on code/modules/client/preferences.dm:
```diff
@@ -177,7 +177,16 @@ GLOBAL_LIST_EMPTY(preferences_datums)
+					dispGender = "Other"
+				dat += "<b>Gender:</b> <a href='?_src_=prefs;preference=gender'>[dispGender]</a><BR>"
+				if(active_character.gender == PLURAL || active_character.gender == NEUTER)
+					dat += "<b>Body Model:</b><a href='?_src_=prefs;preference=body_model'>[active_character.features["body_model"] == MALE ? "Masculine" : "Feminine"]</a><BR>"
```
Personally I don't think hiding the body model pref behind the other gender makes sense. Since the game will use they for the Other gender always, you can't have a character that uses he/him but has the feminine body model or the reverse.

It should certainly default when you change your gender to the matching type but it shouldn't entirely lock you of the option - OR the actual medrecords gender and pronouns should be separated. For now though the previous solution is easier.

Bobbanz1 replied:
> It should certainly default when you change your gender to the matching type but it shouldn't entirely lock you of the option - OR the actual medrecords gender and pronouns should be separated. For now though the previous solution is easier.

Soooo...
Keep it this way or revert it to how it was previously?

PowerfulBacon replied:
There should be a pronoun option in the other category if its not out of scope, but I'm fine with it this way.

itsmeow replied:
The problem is the game makes no distinction between gender and pronouns, as gender functionally is the pronouns. Separating them would be a huge PITA

PowerfulBacon replied:
It would probably need an additional pref which could be a bit of a pain, yea. I'm fine to accept that as something that can just be done in a future PR.
I think for now having the option to have the character as 'other' and then able to choose which body model you want is the best solution.
Alternatively, you could do it entirely with a popup approach, which provides both without making more things to go through for players that use the presets.
![image](https://user-images.githubusercontent.com/26465327/219106852-0655dcd1-2e9c-400b-8e7a-9c45e3feab3c.png)
This would require only a single button for all of them (I guess unless you want to display body model), doesn't need prefs update and does it all via the TGUI

Bobbanz1 committed:
Merge remote-tracking branch 'upstream/master' into AttemptNumberWhatever

Bobbanz1 committed:
More stuff, pulls from downstream some idea from review.

PowerfulBacon commented on code/modules/surgery/healing.dm:
```diff
@@ -70,8 +70,8 @@
-		umsg += " as best as you can while they have clothing on"
-		tmsg += " as best as they can while [target] has clothing on"
+		umsg += " as best as you can while [target.p_they()] [target.p_have()] clothing on"
+		tmsg += " as best as [user.p_they()] can while [target.p_they()] [target.p_have()] clothing on"
```
This doesn't make grammatical sense.
`... as best as they can while they have clothing on`
implies they themselves haev clothing on, rather than the target has clothing on

PowerfulBacon commented on code/modules/mob/living/living.dm:
```diff
@@ -1396,6 +1396,25 @@
+	var/bender = gender != ngender
+	if(forced)
+		gender = ngender
+	if(bender)
```
This doesn't seem to properly account for forced.
Also more appropriate variable names should be used, as oppossed to bender, I think

PowerfulBacon replied:
Should forced not also update the model?

Bobbanz1 replied:
Shifted stuff around, it now updates the model in the new place it does that stuff in.

Bobbanz1 committed:
Reverts the sex field thingy

Bobbanz1 committed:
More updates in response to Reviews

PowerfulBacon commented on code/game/machinery/computer/security.dm:
```diff
@@ -767,12 +767,12 @@ What a mess.*/
-								active1.fields["sex"] = "Female"
-							else
-								active1.fields["sex"] = "Male"
+							var/t1 = input(usr, "Select gender", "Secure. records", active1.fields["gender"]) as null|anything in list(MALE, FEMALE, PLURAL)
```
Won't this say "plural" or does that render as other?

Bobbanz1 replied:
I have fixed this now so it renders as "Other"

PowerfulBacon reviewed (commented):
I think my post in the above comment still applies. I believe that choosing the male pronoun will force the male body type

Bobbanz1 committed:
Update in response to review

itsmeow reviewed (approved):
I don't like the insistence on Male must have a male body type, the preference should be there for any option since you can't functionally separate gender and pronouns, so you can't have a male (he/him) character with a female body type at all currently, you could only have a character with they/them pronouns and a female body type. 

But whatever
